File: engine/SearchPath.java
package com.netadapt.rivalchess.engine;

import com.netadapt.rivalchess.util.ChessBoardConversion;

public class SearchPath 
{
	public int[] move;
	public int score;
	public int height = 0;

	public void reset()
	{
		this.height = 0;
		this.move = null;
		this.score = -RivalConstants.INFINITY;
	}
	
	public void setHeight(int height)
	{
		this.height = height;
	}

	public void setPath(int move)
	{
		this.move = new int[1];
		this.height = 1;
		this.move[0] = move;
	}
	
	public void setPath(SearchPath path)
	{
		int pathLength;
		if (path.move != null)
		{
			pathLength = path.move.length;
			this.move = new int[pathLength];
			for (int i=0; i<pathLength; i++)
			{
	 			this.move[i] = path.move[i];
			}
		}
		
		this.score = path.score;
		this.height = path.height;
	}

	public void setPath(int compactMove, SearchPath path)
	{
		this.height = path.height + 1;
		
		int pathLength;
		if (path.move == null)
		{
			pathLength = 1;
			this.move = new int[pathLength];
		}
		else
		{
			pathLength = path.move.length + 1;
			this.move = new int[pathLength];
		}

		this.move[0] = compactMove;
		
		this.score = path.score;
		
		for (int i=1; i<pathLength; i++)
		{
 			this.move[i] = path.move[i-1];
		}
	}
	
	public String toString()
	{
		String retString = "";
		
		try
		{
			if (this.move != null)
			{
				for (int i=0; i<this.move.length; i++)
				{
					retString += ChessBoardConversion.getSimpleAlgebraicMoveFromCompactMove(this.move[i]) + " ";
				}
			}
		} 
		catch (NullPointerException e)
		{
		}
		
		return retString;
	}
}

File: src/main/java/com/netsensia/rivalchess/engine/core/search/SearchPath.kt
package com.netsensia.rivalchess.engine.core.search

import com.netsensia.rivalchess.config.Limit
import com.netsensia.rivalchess.util.ChessBoardConversion

class SearchPath {
    @JvmField
	val move: IntArray

    @JvmField
    var score = 0
    @JvmField
	var height = 0

    fun reset(): SearchPath {
        height = 0
        score = -Int.MAX_VALUE
        return this
    }

    fun setPath(move: Int) {
        height = 1
        this.move[0] = move
    }

    fun setPath(path: SearchPath) {
        score = path.score
        height = path.height
        if (path.height >= 0) System.arraycopy(path.move, 0, move, 0, path.height)
    }

    fun withScore(newScore: Int): SearchPath {
        score = newScore
        return this
    }

    fun withHeight(newHeight: Int): SearchPath {
        height = newHeight
        return this
    }

    fun withPath(move: Int): SearchPath {
        height = 1
        this.move[0] = move
        return this
    }

    fun withPath(compactMove: Int, path: SearchPath): SearchPath {
        setPath(compactMove, path)
        return this
    }

    fun setPath(compactMove: Int, path: SearchPath) {
        height = path.height + 1
        move[0] = compactMove
        score = path.score
        if (path.height >= 0) System.arraycopy(path.move, 0, move, 1, path.height)
    }

    override fun toString(): String {
        var retString = ""
        try {
            for (i in 0 until height) retString += if (move[i] != 0) ChessBoardConversion.getSimpleAlgebraicMoveFromCompactMove(move[i]) + " " else break
        } catch (e: NullPointerException) {
            throw RuntimeException()
        }
        return retString
    }

    init {
        move = IntArray(Limit.MAX_TREE_DEPTH.value)
    }
}